### DISCUSSION TITLE
How to create a Record in a JS migration

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#1464] [Q&A]
_**This was asked in https://twitter.com/mlusetti/status/1609850246619869184.**_

Currently the JS migrations are not documented well (_I'm planning to update the documentation together with the upcoming JS hooks support in v0.13/v0.14 since it will require restructuring the sections anyway_) but if you want to create a collection or record programmatically you can refer to the Go APIs:
- https://pocketbase.io/docs/collection-methods/
- https://pocketbase.io/docs/record-methods/
- https://pocketbase.io/docs/database/

The main 2 differences are:
- the JS methods are camelCased (eg. `dao.FindRecordById` -> `dao.findRecordById`)
- when a Go method returns an error, the JS version will `throw` instead (aka. errors are not returned as values)

So for example, to create a Record programmatically in a migration you can do something like the following:

```js
// ./pocketbase migrate create 'add_new_record'

migrate((db) => { // up
  const dao = new Dao(db)

  const collection = dao.findCollectionByNameOrId("example")

  const record = new Record(collection, {
    "title": "example",
  })

  dao.saveRecord(record);
}, (db) => { // down
  const dao = new Dao(db)

  // delete the previously created record (if exists)
  try {
    const record = dao.findFirstRecordByData("example", "title", "example")

    dao.deleteRecord(record);
  } catch {}
})
```

**Comment by mlusetti** [comment]
Thank you very much @ganigeorgiev .
I just tweeted this to you and you took time to answer the tweet and file a discussion for me.
Plus the explanation is very well stated with examples too. kudos to you!

**Reply by xrd** [reply]
Thank you so much for the write-up @ganigeorgiev . I've been looking for this and am so excited about the JS migration support.

Can you tell me generally how you would convert an export of the database from the web UI into a JS migration. Should I create another question to ask this?

I'm happy to use the go framework app path if that's easier. I'm trying to seed my database from scratch now that I've got it working on the staging server after lots of experimenting.

**Reply by xrd** [reply]
I'm reading this: https://github.com/pocketbase/pocketbase/issues/48#issuecomment-1321229756

Does this imply that if i use a migration, i need to use pocketbase as a framework app? I'm a little confused about migrations because some of the documentation indicates the migrations are embedded inside the app and I'm using the vanilla binary (inside docker).

**Reply by xrd** [reply]
Wow, I get it. I went into my staging system and looked into pb_migrations. I see a bunch of js files **that were automatically created when I created a collection or made a change**.  All the changes are stored automatically as migrations files. I copied those JS files into the pb_migrations directory, and then when I create a new docker container, the migrations happen automatically. Fantastic.

**Comment by einarpersson** [comment]
@ganigeorgiev 
Hi! New pocketbase user here! :)  I have a DX question related to this very topic.

Are there typings anywhere for the migrations? (`migrate`, `Dao`, `Collection`, ...).

I understand if typescript might not be supported for now (even if that would be _**great**_, from my point of view) but it would be very valuable if the types at least were available somewhere which would make it possible to import them as JSDocs comments and get useful intellisense in the migrations files. This could even be inserted automatically when generating the migration files (suggestion!).

**Reply by ganigeorgiev** [reply]
There are no typings for the JS migrations.

The JS migrations will be documented soon (_the custom queries and aggregations were prioritized for now_) and I'll consider specifying jsdoc comments for some of the types in the future.

I'm not planning to add TypeScript support since it will be hard to maintain. If you want typesafety when writing migrations, my recommendation for now would be to [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and [the Go migrations](https://pocketbase.io/docs/migrations/) (_the automigrate functionality also works with Go templates_).

**Reply by mlusetti** [reply]
May I say I switched to GO migrations (as I already extend Pocketbase) and it works smoothly.

**Reply by einarpersson** [reply]
> There are no typings for the JS migrations.
> 
> The JS migrations will be documented soon (_the custom queries and aggregations were prioritized for now_) and I'll consider specifying jsdoc comments for some of the types in the future.
> 
> I'm not planning to add TypeScript support since it will be hard to maintain. If you want typesafety when writing migrations, my recommendation for now would be to [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and [the Go migrations](https://pocketbase.io/docs/migrations/) (_the automigrate functionality also works with Go templates_).

Providing types in jsdocs would go along way fulfilling my wishes! No need to support running typescript migration files. I found this, I don't know if this is of any use but I put it here in case of: [Generate TypeScript interfaces From Go Structs/interfaces](https://morioh.com/p/2313f3d7dba6)

I understand that you have many priorities and I'm glad to hear that adding _some kind of docs_ for JS migrations is high up on the list. Now I am a bit lost, as a beginner here. 😅 But I guess I'll begin with what you have provided in this thread!

PS. I'm very much looking forward to learn more about Pocketbase and it seems that you are developing something very nice:) Thanks! ⭐

**Reply by einarpersson** [reply]
> The main 2 differences are:
> - the JS methods are camelCased (eg. dao.FindRecordById -> dao.findRecordById)
> - when a Go method returns an error, the JS version will throw instead (aka. errors are not returned as values)

I would definitely prefer if the non-throwing style was present/available also in JS (as `safeParse` in [zod](https://github.com/colinhacks/zod#safeparse) for example) as an option, since using that kind of API forces you to handle errors. Also, when using try-catch you easily end up with a lot of nested try-catches... (but maybe this is a bit off topic! apologize for that!)
